### PR TITLE
Add sys_signalfd/sys_signalfd implementation for container support and signal handling

### PR DESCRIPTION
**Pull request by grief8**
### **Motivation**  
This PR introduces `sys_signalfd` and `sys_signalfd4` system call support, addressing containerization requirements (#703) and enhancing signal handling capabilities. Signalfd is critical for:  
1. **Container Environments**: Isolating signal delivery in namespaces, ensuring signals are handled within the container context.  
2. **Event-Driven Architectures**: Enabling signal reception via file descriptors, simplifying integration with poll/epoll-based event loops.  
3. **Safer Signal Handling**: Avoiding asynchronous signal handlers by treating signals as readable file descriptor events.  

---

### **Feature Overview: signalfd**  
The `signalfd` mechanism allows applications to receive signals via a file descriptor instead of traditional signal handlers.  

**Key Functionality**:  
- **Create/Update Signal FD**:  
  - `sys_signalfd4` creates or updates a file descriptor associated with a signal mask.  
  - Signals in the mask (except `SIGKILL`/`SIGSTOP`) are consumed via `read()`.  
- **Non-Blocking Support**: Configured via `O_NONBLOCK`.  
- **Thread-Safe**: Signals are dequeued atomically.  

**Basic Usage**:  
```c  
sigset_t mask;  
sigemptyset(&mask);  
sigaddset(&mask, SIGINT);  
int fd = signalfd4(-1, &mask, sizeof(sigset_t), O_NONBLOCK | O_CLOEXEC);  

struct signalfd_siginfo info;  
read(fd, &info, sizeof(info));  // Block until signal arrives  
```  

---

### **Implementation Details**  
1. **SignalFile Structure**:  
   - Manages signal mask (via `AtomicSigMask`), observer registration, and blocking/non-blocking I/O.  
   - Integrates with the task's signal queue through the `Observer` pattern.  

2. **Kernel Integration**:  
   - Threads register/unregister signal observers dynamically.  
   - File descriptor operations (`read`, `poll`) trigger signal dequeuing.  

3. **Safety**:  
   - Automatically unregisters observers on `close()`.  
   - Ensures thread lifecycle safety via `Weak<Task>` references.  

---

### **Future Work**  
While this implementation provides core functionality, further improvements are planned:  

1. **Expand siginfo_t Data**:  
   - Current `SignalfdSiginfo` lacks fields like `ssi_pid`, `ssi_uid`, and `ssi_addr`.  

2. **Related System Call Support**:  
> Some givsor syscall test cases are blocked because of the absence of the following syscalls.
   - **`CLONE_SYSVSEM`**: Required for full semaphore support in containerized processes.  
   - **`pidfd`**: Enable signal delivery to processes via PID file descriptors.  
   - **`ppoll`**: Improve timeout precision when polling multiple file descriptors.

### CODE REVIEW

**Comment by grief8** [comment]
## Bug

`Signalfd/SignalfdTest.ThreadGroup/kSigno` of `signalfd_test` fails under different log levels.
```bash
# It fails
make run AUTO_TEST=syscall     SYSCALL_TEST_DIR=/exfat EXTRA_BLOCKLISTS_DIRS=blocklists.exfat     ENABLE_KVM=0 BOOT_PROTOCOL=multiboot2 RELEASE=1 NETDEV=tap 
# It succeeds 
make run AUTO_TEST=syscall     SYSCALL_TEST_DIR=/exfat EXTRA_BLOCKLISTS_DIRS=blocklists.exfat     ENABLE_KVM=0 BOOT_PROTOCOL=multiboot2 RELEASE=1 NETDEV=tap LOG_LEVEL=debug
```
If we only run `signalfd_test`, it still succeeds.

**Comment by grief8** [comment]
@boterinas rerun

File: kernel/src/syscall/signalfd.rs [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+impl FileLike for SignalFile {
+    fn read(&self, writer: &mut VmWriter) -> Result<usize> {
+        if writer.avail() < core::mem::size_of::<SignalfdSiginfo>() {
+            return_errno_with_message!(Errno::EINVAL, "Buffer too small for siginfo structure");
+        }
+
+        if self.is_non_blocking() {
+            self.try_read(writer)
+        } else {
+            self.wait_queue.wait_until(|| Some(self.try_read(writer)))
```
Comment (StevenJiang1110) [inline kernel/src/syscall/signalfd.rs:302]: I'm concerned that this code won't function as expected if the file is blocking. If `try_read` returns `EAGAIN`, the `wait_until` function will simply return `EAGAIN` without actually waiting. You might want to look at `Pollable::wait_events` to reimplement the code correctly.

File: kernel/src/syscall/signalfd.rs [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+    /// Weak reference to associated task
+    task_ref: Weak<Task>,
+    /// Atomic signal mask for filtering signals
+    signals_mask: AtomicSigMask,
+    /// I/O event notifier
+    pollee: Pollee,
+    /// Non-blocking mode flag
+    non_blocking: AtomicBool,
+    /// Observer registration handle
+    observer_handle: Mutex<Option<Weak<dyn Observer<SigEvents>>>>,
```
[the remark this answers is not in the document]
Comment (grief8) [reply kernel/src/syscall/signalfd.rs]: Good idea. It looks better now.

File: kernel/src/syscall/signalfd.rs (hunk 1 above)
Comment (grief8) [reply kernel/src/syscall/signalfd.rs]: Thank you for the advice. It helps simplify the logic.

File: kernel/src/syscall/signalfd.rs [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+            rdev: 0,
+        }
+    }
+}
+
+impl Drop for SignalFile {
+    fn drop(&mut self) {
+        if let Some(thread) = current_thread!().as_posix_thread() {
+            thread.unregister_sigqueue_observer(&self.weak_self);
+        }
```
Comment (StevenJiang1110) [inline kernel/src/syscall/signalfd.rs:319]: This implementation of `drop` seems incorrect. A `SignalFile` can be shared across multiple processes through fork. Consider the scenario where the parent process closes the `SignalFile` first, followed by the child process. In this case, the `drop` function will be triggered when the child closes the file, leading to an attempted unregistration of the observer from the child process. However, since the observer is registered on the parent's signal queue, it will never actually be unregistered.

I think your original design, which involved storing a `Weak<Task>` in `SignalFile`, might still be beneficial. While the `Weak<Task>` may not be useful for reading, it remains useful within the `drop` method.

File: kernel/src/syscall/signalfd.rs [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+        if let Some(thread) = current_thread!().as_posix_thread() {
+            let mask = self.signals_mask.load(Ordering::Relaxed);
+            if thread.sig_pending().intersects(mask) {
+                IoEvents::IN
+            } else {
+                IoEvents::empty()
+            }
+        } else {
+            IoEvents::empty()
+        }
```
Comment (StevenJiang1110) [inline kernel/src/syscall/signalfd.rs:204]: Use `let.. else` to return early to avoid redundant indent.

```rust
let Some(thread) = current_thread!().as_posix_thread() else {
    return IoEvents::empty();
};

let mask = self.signals_mask.load(Ordering::Relaxed);
if thread.sig_pending().intersects(mask) {
    IoEvents::IN
} else {
    IoEvents::empty()
}
```

File: kernel/src/syscall/signalfd.rs [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+    ssi_status: i32,
+    ssi_int: i32,
+    ssi_ptr: u64,
+    ssi_utime: u64,
+    ssi_stime: u64,
+    ssi_addr: u64,
+    _pad: [u8; 48],
+}
+
+trait SignalToSiginfo {
```
Comment (StevenJiang1110) [inline kernel/src/syscall/signalfd.rs:345]: This trait may be better names as `ToSiganlfdSiginfo`. We already has a `siginfo` (siginfo_t), and this trait is of different meaning.

File: kernel/src/syscall/signalfd.rs [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+    struct SignalFileFlags: u32 {
+        const O_CLOEXEC = CreationFlags::O_CLOEXEC.bits();
+        const O_NONBLOCK = StatusFlags::O_NONBLOCK.bits();
+    }
+}
+
+/// Signal file descriptor implementation
+///
+/// Represents a file descriptor that can be used to receive signals
+/// as readable events.
```
Comment (StevenJiang1110) [inline kernel/src/syscall/signalfd.rs:141]: The doc may be better refactored. SignalFile should be a _file_, instead of a _file descripter_.

**Review by StevenJiang1110 — COMMENTED** [review]
Thanks for your efforts! Just a few minor comments.

File: kernel/src/syscall/signalfd.rs (hunk 3 above)
Comment (grief8) [reply kernel/src/syscall/signalfd.rs:323]: Thank you a lot for suggestions! Current implementation indeed has something missed. According to the behavior of Linux, each process should have its own observer, so the `unregister` should also be directed to the process. However, I haven't found an elegant way to detect whether the current process is a child process. 

This issue occurs specifically when a signal is received after the polling operation has completed, causing the blocking `sys_poll` implementation to enter an unrecoverable wait state. I think it can be left as a to-do item.

**Comment by grief8** [comment]
The code below shows the corner case that a signal is received after the polling operation has completed, causing the blocking sys_poll implementation to enter an unrecoverable wait state. 
```c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <sys/signalfd.h>
#include <sys/wait.h>
#include <poll.h>

int main()
{
	sigset_t mask;
	int sfd;
	struct signalfd_siginfo fdsi;
	pid_t pid;

	// Block SIGUSR1
	sigemptyset(&mask);
	sigaddset(&mask, SIGUSR1);
	if (sigprocmask(SIG_BLOCK, &mask, NULL) == -1) {
		perror("sigprocmask");
		exit(EXIT_FAILURE);
	}

	// Create signalfd
	sfd = signalfd(-1, &mask, 0);
	if (sfd == -1) {
		perror("signalfd");
		exit(EXIT_FAILURE);
	}

	// Create child process
	pid = fork();
	if (pid == -1) {
		perror("fork");
		exit(EXIT_FAILURE);
	}

	if (pid == 0) {
		// Child process logic
		printf("Child PID: %d\n", getpid());

		// Verify signal mask
		sigset_t current_mask;
		sigprocmask(SIG_BLOCK, NULL, &current_mask);
		if (sigismember(&current_mask, SIGUSR1)) {
			printf("Child: SIGUSR1 is blocked.\n");
		} else {
			printf("Child: SIGUSR1 is NOT blocked!\n");
			exit(EXIT_FAILURE);
		}

		sleep(1);

		// // Child sends signal to self
		// printf("Child sending SIGUSR1 to self...\n");
		// kill(getpid(), SIGUSR1);

		// Monitor for signal
		struct pollfd pfd = { .fd = sfd, .events = POLLIN };
		int ret = poll(&pfd, 1, 5000);
		if (ret == -1) {
			perror("poll");
			exit(EXIT_FAILURE);
		} else if (ret == 0) {
			printf("Child: poll timeout (no signal).\n");
		} else {
			if (pfd.revents & POLLIN) {
				read(sfd, &fdsi, sizeof(fdsi));
				printf("Child received SIGUSR1 (sender PID: %d)\n",
				       fdsi.ssi_pid);
			}
		}

		close(sfd);
		exit(EXIT_SUCCESS);
	} else {
		// Parent process logic
		printf("Parent PID: %d\n", getpid());
		printf("Parent closing fd: %d\n", sfd);
		close(sfd);

		// Parent sends signal to child
		sleep(1); // Wait for child to enter poll state
		printf("Parent sending SIGUSR1 to child...\n");
		kill(pid, SIGUSR1);

		// Wait for child to exit
		wait(NULL);
		exit(EXIT_SUCCESS);
	}
}
```

**Comment by StevenJiang1110** [comment]
> The code below shows the corner case that a signal is received after the polling operation has completed, causing the blocking sys_poll implementation to enter an unrecoverable wait state.

Thank you for your test. It suggests that we might need to inherit the sigqueue observer, even though the sigqueue is newly created, when cloning a child process in certain cases. However, implementing this could introduce additional complexity. It's fine to leave this as future work.

File: kernel/src/syscall/signalfd.rs (hunk 3 above)
Comment (StevenJiang1110) [reply kernel/src/syscall/signalfd.rs:323]: It may be better to also add a FIXME in the drop method to note that possible resource leakage may happen.

**Comment by grief8** [comment]
@boterinas ready

File: kernel/src/process/signal/sig_mask.rs [hunk 7]
```diff
@@ -143,6 +151,11 @@ impl SigSet {
         let set = set.into();
         self.bits & set.bits == set.bits
     }
+
+    pub fn intersects(&self, set: impl Into<Self>) -> bool {
```
Comment (tatetian) [inline kernel/src/process/signal/sig_mask.rs:155]: Rename `set` to `other`. This conforms with the naming convention of Rust std.

Rename this method to `does_intersect` to make it more obvious that the method returns a bool. Otherwise, people may think that this method returns the intersection of `self` and `other`.

File: kernel/src/syscall/signalfd.rs [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+            sig_mask::{AtomicSigMask, SigMask},
+            signals::Signal,
+            PollHandle, Pollable, Pollee, SigEvents, SigEventsFilter,
+        },
+        Gid, Uid,
+    },
+    time::clocks::RealTimeClock,
+};
+
+/// Create new signalfd file descriptor
```
Comment (tatetian) [inline kernel/src/syscall/signalfd.rs:35]:
```
/// Creates a new signalfd or updates an existing one according to the given mask
```

File: kernel/src/syscall/signalfd.rs [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+pub fn sys_signalfd(
+    fd: FileDesc,
+    mask_ptr: Vaddr,
+    sizemask: usize,
+    ctx: &Context,
+) -> Result<SyscallReturn> {
+    sys_signalfd4(fd, mask_ptr, sizemask, 0, ctx)
+}
+
+/// Create/update signalfd with flags
```
Comment (tatetian) [inline kernel/src/syscall/signalfd.rs:45]:
```
/// Creates a new signalfd or updates an existing one according to the given mask and flags
```

File: kernel/src/syscall/signalfd.rs [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+            IoEvents::IN
+        } else {
+            IoEvents::empty()
+        }
+    }
+
+    /// Attempt non-blocking read operation
+    fn try_read(&self, writer: &mut VmWriter) -> Result<usize> {
+        let buffer_size = writer.avail();
+        if buffer_size % core::mem::size_of::<SignalfdSiginfo>() != 0 {
```
Comment (tatetian) [inline kernel/src/syscall/signalfd.rs:211]: Are you sure that this is a requirement? I really don't think so.

File: kernel/src/syscall/signalfd.rs [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+            return_errno_with_message!(
+                Errno::EINVAL,
+                "Buffer size must be multiple of SignalfdSiginfo size"
+            );
+        }
+
+        let binding = current_thread!();
```
Comment (tatetian) [inline kernel/src/syscall/signalfd.rs:218]: You need to follow the naming convention of the existing code. This is a large codebase where many developers collaborate. If everyone feels free to invent his own style, the codebase would be a mess. Check out existing user code of `current_thread!` and see how they name the obtained object.

Make sure that all appearances of `binding` are fixed.

**Review by tatetian — COMMENTED** [review]
The code looks great. I raised some tiny code style issues.

File: kernel/src/process/signal/sig_mask.rs (hunk 7 above)
Comment (grief8) [reply kernel/src/process/signal/sig_mask.rs:155]: Thank you for your suggestions! Renaming `set` to `other` is a better choice.

> Rename this method to does_intersect to make it more obvious that the method returns a bool. 

I think `intersects` is enough. It aligns with the conventions of `bitflags`, which is widely used in Asterinas, e.g., `containers`, `intersects`.

File: kernel/src/syscall/signalfd.rs (hunk 10 above)
Comment (grief8) [reply kernel/src/syscall/signalfd.rs:211]: You are right. The check of `read` is enough.

File: kernel/src/syscall/signalfd.rs (hunk 11 above)
Comment (grief8) [reply kernel/src/syscall/signalfd.rs]: Thank you for your suggestion. I have renamed it to `current` now, which aligns with the following one.
https://github.com/asterinas/asterinas/blob/dd97f016a32f6f51089ea81239037580e44dbd2f/kernel/src/net/socket/unix/ns/path.rs#L15-L16

**Comment by grief8** [comment]
@boterinas rerun

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!
